### PR TITLE
Bug fix in emcee hammer

### PR DESCRIPTION
**Pull request by ben18785**
- Separated sample_z out into its own function as this makes more explicit what's happening
- Cleaned up choosing a member from the complementary set
- Corrected the setter for scale (and corrected its test)

Addresses #1325

### CODE REVIEW

File: pints/_mcmc/_emcee_hammer.py [hunk 1]
```diff
@@ -82,15 +82,13 @@ def ask(self):
             self._remaining = np.delete(self._remaining, 0)
 
             # Pick j from the complementary ensemble
-            j = np.random.randint(self._n_chains - 1)
-            if j >= self._k:
-                j += 1
+            compl_indices = list(set(range(self._n_chains)) -
+                                 set([self._k]))
+            j = np.random.choice(compl_indices)
```
Comment (DavAug) [inline pints/_mcmc/_emcee_hammer.py:87]: Sorry I don't know much about this sampler. 
- k is the index of the current runner/chain?
- j is the new runner that you want to pick?

Would it potentially be easier (and at least for me easier to understand) if you did something like:
```
# Create list of runners (Could also be done in the init)
self._runners = list(range(self._n_chains)) 

# Pick free runner
free_runners = self._runners.copy()
free_runners.remove(self._k)
j = np.random.choice(free_runners)
```

File: pints/_mcmc/_emcee_hammer.py [hunk 2]
```diff
@@ -101,6 +99,14 @@ def ask(self):
         # Return proposed points
         return self._proposed
 
+    def _sample_z(self, a):
+        """
+        Samples ``z~g(z)`` where ``g(z)`` is proportional to ``1 / sqrt(z)``
+        if ``z`` is in ``[1 / a, a]``; otherwise 0.
+        """
+        r = np.random.rand()
+        return ((1 + r * (a - 1))**2) / a
```
Comment (DavAug) [inline pints/_mcmc/_emcee_hammer.py:108]: Could you explain how this formula relates to g(z) propto 1/sqrt(z) and the interval condition on z?

File: pints/_mcmc/_emcee_hammer.py [hunk 3]
```diff
@@ -160,7 +166,7 @@ def tell(self, fx):
         accepted = np.array([False] * self._n_chains)
         r_log = np.log(np.random.rand())
         q = (
-            (self._n_chains - 1) * np.log(self._z)
+            (self._n_parameters - 1) * np.log(self._z)
             + proposed_log_pdf[0] - self._current_log_pdfs[self._k])
```
Comment (DavAug) [inline pints/_mcmc/_emcee_hammer.py:170]: Could you explain this q to me? Is this the log q of the q mentioned in the docstring
```
q = z^{N - 1} p(Y) / p(X_k(t))?
```
If so, maybe `log_q` would be more clear? What do you think?

In the docstring it also says: ``N`` is the number of chains (or walkers). So do you really want to use the number of parameters, or is the docstring wrong?

**Review by DavAug — CHANGES_REQUESTED** [review]
Thanks @ben18785 !

I don't know much about the sampler, so I apologise for potentially uninformed comments! It looks good though, there are just some bits that are not immediately clear to me.

File: pints/_mcmc/_emcee_hammer.py (hunk 1 above)
Comment (ben18785) [reply pints/_mcmc/_emcee_hammer.py:87]: Good point: I've created another method that handles this and uses the code you've posted. Re: k and j, yes, that's right: I know j and k aren't informative but they match the notation in the paper, so I'd like to keep them like that.

File: pints/_mcmc/_emcee_hammer.py (hunk 2 above)
Comment (ben18785) [reply pints/_mcmc/_emcee_hammer.py:108]: I've added a note to indicate that the method uses inverse transform sampling. I'm not sure I want to explain the formula itself though...

File: pints/_mcmc/_emcee_hammer.py (hunk 3 above)
Comment (ben18785) [reply pints/_mcmc/_emcee_hammer.py:170]: Renamed to `log_q`: good idea! And, yes, good spot in the docstrings: they were wrong and have now been corrected.

**Comment by ben18785** [comment]
Thanks for the review @DavAug -- I think I've changed everything as you wanted. Let me know if you're now happy for this to go in.

**Review by DavAug — APPROVED** [review]
Thanks @ben18785 . It looks really good to me!

**Comment by ben18785** [comment]
Thanks @DavAug !

**Pull request merged**